core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/BufferingAsyncRequestBody.java

```diff
@@ -0,0 +1,322 @@
+ *   <li>Each subscriber receives a read-only view of the buffered data</li>
+ *   <li>Subscribers receive data independently based on their own demand signaling</li>
+ *   <li>If the body is closed, new subscribers will receive an error immediately</li>
+ * </ul>
+ * 
+ * <h3>Resource Management:</h3>
+ * The body should be closed when no longer needed to free buffered data and notify active subscribers.
+ * Closing the body will:
+ * <ul>
+ *   <li>Clear all buffered data</li>
+ *   <li>Send error notifications to all active subscribers</li>
+ *   <li>Prevent new subscriptions</li>
+ * </ul>
+ *
+ */
+@ThreadSafe
+@SdkInternalApi
+public final class BufferingAsyncRequestBody implements AsyncRequestBody, SdkAutoCloseable {
+    private static final Logger log = Logger.loggerFor(BufferingAsyncRequestBody.class);
+
+    private final Long length;
+    private final List<ByteBuffer> bufferedData = new ArrayList<>();
+    private boolean dataReady;
+    private boolean closed;
+    private final Set<ReplayableByteBufferSubscription> subscriptions;
+    private final Object lock = new Object();
+
+    /**
+     * Creates a new BufferingAsyncRequestBody with the specified content length.
+     * 
+     * @param length the total content length in bytes, or null if unknown
+     */
+    BufferingAsyncRequestBody(Long length) {
+        this.length = length;
+        this.subscriptions = ConcurrentHashMap.newKeySet();
+    }
+
+    /**
+     * Sends a chunk of data to be buffered for later consumption by subscribers.
+     * 
+     * @param data the data to buffer, must not be null
+     * @throws NullPointerException if data is null
+     */
+    public void send(ByteBuffer data) {
+        Validate.paramNotNull(data, "data");
+        synchronized (lock) {
+            if (closed) {
+                throw new IllegalStateException("Cannot send data to closed body");
+            }
+            if (dataReady) {
+                throw new IllegalStateException("Request body has already been completed");
+            }
+            bufferedData.add(data);
+        }
+    }
+
+    /**
+     * Marks the request body as complete and ready for subscription.
+     * 
+     * <p>This method must be called before any subscribers can successfully subscribe
+     * to this request body. After calling this method, no more data should be sent
+     * via {@link #send(ByteBuffer)}.
+     * 
+     * <p>Once complete, multiple subscribers can subscribe and will each receive
+     * the complete buffered content independently.
+     */
+    public void complete() {
+        synchronized (lock) {
+            if (dataReady) {
+                return;
+            }
+            if (closed) {
+                throw new IllegalStateException("The AsyncRequestBody has been closed");
+            }
+            dataReady = true;
+        }
+    }
+
+    @Override
+    public Optional<Long> contentLength() {
+        return Optional.ofNullable(length);
```

Comment:
It's kind of weird to me that `contentLength` could return an empty optional, but we require all the data to be buffered first before we start publishing. Should we just require all of the data to be ready at construction time?

Reply:
Hmm, good point!  It'll also reduce the complexity